File: lib/wDispatcher/dispatcher.cpp
#include "dispatcher.h"

W::Dispatcher::Dispatcher()
{}

W::Dispatcher::~Dispatcher()
{}

void W::Dispatcher::pass(const W::Event& ev) const
{

    n_map::const_iterator itr;
    itr = nodes.find(ev.getDestination());
        
    if (itr != nodes.end())
    {
        W::Order<W::Handler*>::const_iterator beg = itr->second.begin();
        W::Order<W::Handler*>::const_iterator end = itr->second.end();
        
        std::list<W::Handler*> list(beg, end);
        std::list<W::Handler*>::const_iterator itr = list.begin();
        std::list<W::Handler*>::const_iterator tEnd = list.end();
        for (; itr != tEnd; ++itr) {
            (*itr)->pass(ev);
        }
    }
    else
    {
        d_order::const_iterator itr = defaultHandlers.begin();
        d_order::const_iterator end = defaultHandlers.end();
        
        for (; itr != end; ++itr) 
        {
            if ((*itr)->call(ev)){
                break;
            }
        }
    }
}

void W::Dispatcher::registerHandler(W::Handler* dp)
{
    n_map::iterator itr = nodes.find(dp->getAddress());
    if (itr == nodes.end()) {
        W::Order<W::Handler*> ord;
        itr = nodes.insert(std::make_pair(dp->getAddress(), ord)).first;
    }
    itr->second.push_back(dp);
}

void W::Dispatcher::unregisterHandler(W::Handler* dp)
{
    n_map::iterator itr = nodes.find(dp->getAddress());
    if (itr != nodes.end())
    {
        W::Order<W::Handler*>::const_iterator o_itr = itr->second.find(dp);
        if (o_itr != itr->second.end()) {
            itr->second.erase(dp);
        
            if (itr->second.size() == 0) {
                nodes.erase(itr);
            }
        }
        else
        {
            throw 5;//TODO exception;
        }
        
    }
    else
    {
        throw 5;//TODO exception;
    }
}

void W::Dispatcher::registerDefaultHandler(W::DefaultHandler* dh)
{
    defaultHandlers.push_back(dh);
}

void W::Dispatcher::unregisterDefaultHandler(W::DefaultHandler* dh)
{
    defaultHandlers.erase(dh);
}

File: libjs/wDispatcher/dispatcher.js
"use strict";
var Class = require("../utils/class");
var Handler = require("./handler");
var DefaultHandler = require("./defaulthandler");
var Address = require("../wType/address");
var AbstractMap = require("../wContainer/abstractmap");
var AbstractOrder = require("../wContainer/abstractorder");

var HandlerOrder = AbstractOrder.template(Handler);
var DefaultHandlerOrder = AbstractOrder.template(DefaultHandler);
var HandlerMap = AbstractMap.template(Address, HandlerOrder);

var Dispatcher = Class.inherit({
    "className": "Dispatcher",
    "constructor": function() {
        Class.fn.constructor.call(this);
        
        this._handlers = new HandlerMap();
        this._defautHandlers = new DefaultHandlerOrder(false);
    },
    "destructor": function() {
        this._handlers.destructor();
        
        Class.fn.destructor.call(this);
    },
    "registerHandler": function(handler) {
        var itr = this._handlers.find(handler.address);
        var order;
        
        
        if (itr["=="](this._handlers.end())) {
            order = new HandlerOrder(false);
            this._handlers.insert(handler.address.clone(), order);
        } else {
            order = itr["*"]();
        }
        
        order.push_back(handler);
    },
    "unregisterHandler": function(handler) {
        var itr = this._handlers.find(handler.address);
        
        if (!itr["=="](this._handlers.end())) {
            
            var ord = itr["*"]().second;
            ord.erase(handler);
            
            if (ord.size() === 0) {
                this._handlers.erase(itr);
            }
            
        } else {
            throw new Error("Can't unregister hander");
        }
    },
    "registerDefaultHandler": function(dh) {
        this._defautHandlers.push_back(dh);
    },
    "unregisterDefaultHandler": function(dh) {
        this._defautHandlers.erase(dh);
    },
    "pass": function(event) {
        var itr = this._handlers.find(event.getDestination());
        
        if (!itr["=="](this._handlers.end())) {
            var ord = itr["*"]().second;
            
            var o_beg = ord.begin();
            var o_end = ord.end();
            
            for (; !o_beg["=="](o_end); o_beg["++"]()) {
                o_beg["*"]().pass(event);
            }
        } else {
            var dhitr = this._defautHandlers.begin();
            var dhend = this._defautHandlers.end();
            for (; !dhitr["=="](dhend); dhitr["++"]()) {
                dhitr["*"]().call(event);
            }
        }
    }
});

module.exports = Dispatcher;
